### PR TITLE
docs: How to make a contract with a multiple attribute

### PR DESCRIPTION
**Pull request by jeromecambon**
Covers [BPM-321](https://bonitasoft.atlassian.net/browse/BPM-321)

[BPM-321]: https://bonitasoft.atlassian.net/browse/BPM-321?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: modules/applications/pages/ui-builder/how-to-contract-with-multiple-attribute.adoc [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+{description}
+
+* Difficulty: intermediate
+* Estimated time: 20 minutes
+
+With Bonita, you can create a contract with a `multiple` attribute.
+For example, a provider may have multiple contacts.
+
+Creating a UI to start a Process Instance, you will need a UI interface that allows the user to create a new Provider and add multiple Contacts.
```
Comment (marcobonita) [inline modules/applications/pages/ui-builder/how-to-contract-with-multiple-attribute.adoc]:
```suggestion
To start a Process Instance, the UI should allow users to create a new Provider and add multiple Contacts.
```

**Comment by AleRondon** [comment]
Feedback after using the described procedure in [Invoice Management App](https://github.com/bonitasoft-presales/uib_InvoiceManagement/)

1. **Impossible to start with no contact**: The described documentation requires to start with an existing contact / placeholder, which might be confusing for some users. Should be able to have an empty array and create a new line with the required attributes.
2. **Update function**: Updating a contact just erases and create a new one which can lead to next issue.
3. **ID management errors**: ID can be duplicated if deleting a contact and recreating a new one, which can then cause errors when deleting or updating a contact.

image

File: modules/applications/pages/ui-builder/how-to-contract-with-multiple-attribute.adoc [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+In this guide, we will focus on how to create a list of contacts, and store them in a JSON object.
+This will be used to fill the `contacts` attribute in the contract.
+
+== Create a JS Object to manage the list of contacts
+
+Go to the `JS` tab and create a new JS Object called `contactsHandler` and copy and paste the following code:
```
Comment (marcobonita) [inline modules/applications/pages/ui-builder/how-to-contract-with-multiple-attribute.adoc]:
```suggestion
Go to the `JS` tab and create a new JS Object called `ContactsHandler` and copy and paste the following code:
```

File: modules/applications/pages/ui-builder/how-to-contract-with-multiple-attribute.adoc [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    ** Set the `onSave` property to `Execute a JS function` -> `ContactsHandler.add(data)` with `{{Table1.newRow}}` parameter:
+
+image:ui-builder/guides/adding-a-row-on-save.png[adding-a-row-on-save, 500]
+
+* From the `Columns` section, click on the `Save / Discard` spin wheel icon to edit the column properties
+* Set the `onSave` property as `JS` and as `{{ContactsHandler.update(Table1.updatedRow)}}` :
+
+image:ui-builder/guides/table-save-discard-column.png[table-save-discard-column, 300]
+
+* Set the `Title` column as `Select` and set the `Options` property as `{{ContactsHandler.getTitleTypes.data}}`
```
Comment (marcobonita) [inline modules/applications/pages/ui-builder/how-to-contract-with-multiple-attribute.adoc]:
```suggestion
* Set the `Title` column type as `Select` and set the `Options` property as `{{ContactsHandler.getTitleTypes.data}}`
```

File: modules/applications/pages/ui-builder/how-to-contract-with-multiple-attribute.adoc [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+* Set the `ContactType` column as `Select` and set the `Options` property as `{{ContactsHandler.getContactTypes.data}}`
```
Comment (marcobonita) [inline modules/applications/pages/ui-builder/how-to-contract-with-multiple-attribute.adoc]:
```suggestion
* Set the `ContactType` column type as `Select` and set the `Options` property as `{{ContactsHandler.getContactTypes.data}}`
```

File: modules/applications/pages/ui-builder/how-to-contract-with-multiple-attribute.adoc [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+=== Add a button to delete a row
+
+* Add a `delete` column to the table
+* Edit the column properties:
+    ** Set the `Column type` as `delete` icon
```
Comment (marcobonita) [inline modules/applications/pages/ui-builder/how-to-contract-with-multiple-attribute.adoc]:
```suggestion
    ** Set the `Column type` as `icon button`
```

**Review by marcobonita — CHANGES_REQUESTED** [review]

**Comment by jeromecambon** [comment]
> Feedback after using the described procedure in [Invoice Management App](https://github.com/bonitasoft-presales/uib_InvoiceManagement/)
> 
> 1. **Impossible to start with no contact**: The described documentation requires to start with an existing contact / placeholder, which might be confusing for some users. Should be able to have an empty array and create a new line with the required attributes.
> 2. **Update function**: Updating a contact just erases and create a new one which can lead to next issue.
> 3. **ID management errors**: ID can be duplicated if deleting a contact and recreating a new one, which can then cause errors when deleting or updating a contact.
> 
> image



> Feedback after using the described procedure in [Invoice Management App](https://github.com/bonitasoft-presales/uib_InvoiceManagement/)
> 
> 1. **Impossible to start with no contact**: The described documentation requires to start with an existing contact / placeholder, which might be confusing for some users. Should be able to have an empty array and create a new line with the required attributes.
> 2. **Update function**: Updating a contact just erases and create a new one which can lead to next issue.
> 3. **ID management errors**: ID can be duplicated if deleting a contact and recreating a new one, which can then cause errors when deleting or updating a contact.
> 
> image

I have updated the JS code to make it work correctly, without initial contact.

**Comment by marcobonita** [comment]
@jeromecambon could we elaborate a little bit on the purpose of this guide? 

For example the introduction "With Bonita, you can create a contract with a multiple attribute. For example, a provider may have multiple contacts." could use a little bit more details. It doesn't seem really clear to me for now

**Comment by AleRondon** [comment]
The ID issues (to create, update and delete users) has been tested and fixed on my application
For the empty contact, maybe it will be best to leave an example/placeholder, to avoid having an empty table without columns or any other strange behaviour

**Review by marcobonita — APPROVED** [review]

**Review by benjaminParisel — APPROVED** [review]

**Pull request merged**